File: ch04/26/Clone.cpp
/*
struct RandomListNode {
    int label;
    struct RandomListNode *next, *random;
    RandomListNode(int x) :
            label(x), next(NULL), random(NULL) {
    }
};
*/
class Solution {
public:
    //第二遍刷题
    RandomListNode* Clone(RandomListNode* pHead)
    {
        if (pHead == NULL)
            return NULL;
        
        //先把复制的新节点插入到原链表中
        RandomListNode* tmpNode = pHead;
        while (tmpNode != NULL) {
            RandomListNode *newNode = new RandomListNode(tmpNode->label);
            newNode->next = tmpNode->next;
            tmpNode->next = newNode;
            
            tmpNode = newNode->next;
        }
        
        //为新节点的random指针赋值
        tmpNode = pHead;
        while (tmpNode != NULL) {
            RandomListNode* newNode = tmpNode->next;
            if (tmpNode->random != NULL)
                newNode->random = tmpNode->random->next;

            tmpNode = newNode->next;
        }
        
        //把链表拆分成两个链表
        tmpNode = pHead;
        RandomListNode* newHead = pHead->next;
        RandomListNode* newNode = newHead;        
        tmpNode->next = newNode->next;
        tmpNode = tmpNode->next;
        while (tmpNode != NULL) {
            newNode->next = tmpNode->next;
            newNode = newNode->next;
            tmpNode->next = newNode->next;
            tmpNode = tmpNode->next;
        }
        
        return newHead;
    }
};

File: ch04/26/Clone.py
# -*- coding:utf-8 -*-
class RandomListNode:
    def __init__(self, x):
        self.label = x
        self.next = None
        self.random = None

class Solution:
    # 返回 RandomListNode
    def Clone(self, pHead):
        # write code here
        if pHead == None:
            return None

        p = pHead
        while p:
            newNode = RandomListNode(p.label)
            newNode.next = p.next
            p.next = newNode
            p = newNode.next

        p = pHead
        # newP = p.next
        while p:
            newP = p.next
            if p.random:
                newP.random = p.random.next
            p = newP.next
        
        p = pHead
        newP = pHead.next
        newHead = pHead.next
        while newP.next:
            p.next = newP.next
            p = p.next
            newP.next = p.next
            newP = newP.next
        p.next = None

        return newHead

if __name__ == "__main__":
    # {1,2,3,4,5,3,5}
    n1 = RandomListNode(1)
    n2 = RandomListNode(2)
    n3 = RandomListNode(3)
    n4 = RandomListNode(4)
    n5 = RandomListNode(5)
    n1.next = n2
    n1.random = n3
    n2.next = n3
    n3.next = n4
    n4.next = n5
    n4.random = n2

    n = n1
    print "row list:"
    while n:
        print n.label
        n = n.next

    s = Solution()
    copy = s.Clone(n1)
    tmp = copy
    print "copy list:"
    while (tmp):
        print tmp.label
        tmp = tmp.next